ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/State/Infra.hs

```diff
@@ -170,56 +170,84 @@ reconstructSummary (History.Shape shape) transition (HardForkState st) =
     go :: Exactly xs' EraParams
```

Comment:
This PR is draft because I think the new code below is correct, but I'm in the process of convincing myself (and _en passant_ writing missing documentation). In particular, I need to understand exactly when `TransitionKnown`, `TransitionUnknown`, and in particular `TransitionImpossible` both _can_ and _cannot_ arise.

This code dates back to Edsko's original PR #2034 two years ago, and this code unfortunately didn't get any discussion that archived on that PR (Thomas and Edsko would review on Meet and write-down TODOs -- it's possible they talked about it without writing a comment on the PR).

The other aspect of due diligence is to catalog the downstream consequences of this change. Which is unfortunately kind of broad -- for those with access, you can see the use-def tree here https://input-output.atlassian.net/browse/CAD-4296?focusedCommentId=91616